<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="uniprotREST">
<title>Getting started with uniprotREST • uniprotREST</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with uniprotREST">
<meta property="og:description" content="uniprotREST">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">uniprotREST</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/uniprotREST.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/formats.html">Available Formats</a>
    <a class="dropdown-item" href="../articles/map_dbs.html">Map Databases</a>
    <a class="dropdown-item" href="../articles/return_fields_other.html">Return Fields - Other Databases</a>
    <a class="dropdown-item" href="../articles/return_fields_uniprot.html">Return Fields - UniProtKB</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting started with uniprotREST</h1>
            
      
      
      <div class="d-none name"><code>uniprotREST.Rmd</code></div>
    </div>

    
    
<p>This document will show you the basics of
<strong>uniprotREST</strong>. This package uses <a href="https://httr2.r-lib.org" class="external-link">httr2</a> to wrap the latest UniProt REST
API, which was updated in June 2022. I wrote this package as an
easy-to-use interface to the API for R users who need to regularly and
reproducibly download information from UniProt.</p>
<p><strong>uniprotREST</strong> has 3 main functions to use:</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/uniprot_map.html">uniprot_map()</a></code> to map <em>to</em> or <em>from</em>
UniProt accessions.</li>
<li>
<code><a href="../reference/uniprot_search.html">uniprot_search()</a></code> to perform text search queries.</li>
<li>
<code><a href="../reference/uniprot_single.html">uniprot_single()</a></code> to get detailed information for a
single entry.</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">uniprotREST</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="id-mapping-with-uniprot_map">1. ID mapping with <code>uniprot_map</code><a class="anchor" aria-label="anchor" href="#id-mapping-with-uniprot_map"></a>
</h2>
<p>This is by far the most frequently used tool. Say hypothetically, you
have been given a list of <a href="https://www.uniprot.org/help/accession_numbers" class="external-link">UniProt
accessions</a>. You have no clue what proteins they refer to, or what
properties these proteins have. You can use <code><a href="../reference/uniprot_map.html">uniprot_map()</a></code>
to find this out.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Accessions of interest</span></span>
<span><span class="va">aoi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A0A8I6AN81"</span>, <span class="st">"A0A0N4SVP8"</span>, <span class="st">"Q9H6R0"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="default-settings">Default settings<a class="anchor" aria-label="anchor" href="#default-settings"></a>
</h3>
<p>Here we just use the default settings, which will map the IDs from
<code>UniProtKB_AC-ID</code> to <code>UniProtKB</code>, and output a
dataframe.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>result1 <span class="ot">&lt;-</span> <span class="fu">uniprot_map</span>(<span class="at">ids =</span> aoi)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Running job: aa29fe3eb9b54ec1e3bff6890465804ef3085fc6 </span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a> Downloading<span class="sc">:</span> page <span class="dv">1</span> of <span class="dv">1</span></span></code></pre></div>
<p>The job ID is automatically printed (stop printing by setting the
<code>verbosity</code> argument to 0). Job IDs and the job data are kept
by UniProt for approximately 7 days, and are then deleted.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># All 3 proteins are RNA helicases</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result1</span><span class="op">)</span></span>
<span><span class="co">##         From      Entry       Entry.Name   Reviewed</span></span>
<span><span class="co">## 1 A0A8I6AN81 A0A8I6AN81   A0A8I6AN81_RAT unreviewed</span></span>
<span><span class="co">## 2 A0A0N4SVP8 A0A0N4SVP8 A0A0N4SVP8_MOUSE unreviewed</span></span>
<span><span class="co">## 3     Q9H6R0     Q9H6R0      DHX33_HUMAN   reviewed</span></span>
<span><span class="co">##                                                          Protein.names</span></span>
<span><span class="co">## 1                                           RNA helicase (EC 3.6.4.13)</span></span>
<span><span class="co">## 2                                           RNA helicase (EC 3.6.4.13)</span></span>
<span><span class="co">## 3 ATP-dependent RNA helicase DHX33 (EC 3.6.4.13) (DEAH box protein 33)</span></span>
<span><span class="co">##    Gene.Names                Organism Length</span></span>
<span><span class="co">## 1  Rig1 Ddx58 Rattus norvegicus (Rat)    881</span></span>
<span><span class="co">## 2    Eif4a3l2    Mus musculus (Mouse)    411</span></span>
<span><span class="co">## 3 DHX33 DDX33    Homo sapiens (Human)    707</span></span></code></pre></div>
<p>By default, the output will be a dataframe with 8 columns:</p>
<ul>
<li>
<code>From</code> = accessions used to map from</li>
<li>
<code>To</code> = accessions they were mapped to</li>
<li>
<code>Entry.Name</code> = UniProtKB entry name</li>
<li>
<code>Reviewed</code> = is the protein in Swiss-Prot?</li>
<li>
<code>Protein.names</code> = name of protein in UniProtKB</li>
<li>
<code>Gene.Names</code> = gene names associated with this protein
(can be multiple)</li>
<li>
<code>Organism</code> = name of organism the protein is from</li>
<li>
<code>Length</code> = amino acid length</li>
</ul>
<p>And n rows which depends on:</p>
<ul>
<li>How many ids were successfully mapped</li>
<li>If the mapping was 1:1 or not</li>
</ul>
<p>The output columns can be customised with the <code>fields</code>
argument.</p>
</div>
<div class="section level3">
<h3 id="return-fields">Return fields<a class="anchor" aria-label="anchor" href="#return-fields"></a>
</h3>
<p>UniProt has a lot of metadata available for each protein. You can
access this in the results by requesting different columns or ‘return
fields’ using the <code>fields</code> argument.</p>
<p>Here we will request some different return fields. See the X page for
the list of all available <code>fields</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Jobs are stored for 7 days so this time</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># it will be quicker!</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>result2 <span class="ot">&lt;-</span> <span class="fu">uniprot_map</span>(</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">ids =</span> aoi,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">fields =</span> <span class="fu">c</span>(</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gene_primary"</span>,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"organism_name"</span>,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"length"</span>,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"mass"</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Running job: aa29fe3eb9b54ec1e3bff6890465804ef3085fc6 </span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a> Downloading<span class="sc">:</span> page <span class="dv">1</span> of <span class="dv">1</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result2</span><span class="op">)</span></span>
<span><span class="co">##         From Gene.Names..primary.                Organism Length   Mass</span></span>
<span><span class="co">## 1 A0A8I6AN81                 Rig1 Rattus norvegicus (Rat)    881 101151</span></span>
<span><span class="co">## 2 A0A0N4SVP8             Eif4a3l2    Mus musculus (Mouse)    411  46959</span></span>
<span><span class="co">## 3     Q9H6R0                DHX33    Homo sapiens (Human)    707  78874</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fromto-database">From/to database<a class="anchor" aria-label="anchor" href="#fromto-database"></a>
</h3>
<p><code><a href="../reference/uniprot_map.html">uniprot_map()</a></code> can be used to map IDs from other
databases to UniProt IDs, and vice-versa. See the X page for a list of
databases available for mapping, and X page for the rules of which
databases can be mapped to what.</p>
<p>Here we’ll map some Ensembl gene IDs to <em>reviewed</em> UniProtKB
accessions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Genes of interest</span></span>
<span><span class="va">goi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000088247"</span>, <span class="st">"ENSG00000162613"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>fields</code> argument only works when mapping <em>to</em>
a UniProtKB, UniRef, or UniParc database.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>result3 <span class="ot">&lt;-</span> <span class="fu">uniprot_map</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">ids =</span> goi,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"Ensembl"</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"UniProtKB-Swiss-Prot"</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">fields =</span> <span class="fu">c</span>(<span class="st">"accession"</span>, <span class="st">"gene_primary"</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Running job: 03772c110ee17372f3dc29049d1ff9fa50105fad </span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a> Downloading<span class="sc">:</span> page <span class="dv">1</span> of <span class="dv">1</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result3</span><span class="op">)</span></span>
<span><span class="co">##              From  Entry Gene.Names..primary.</span></span>
<span><span class="co">## 1 ENSG00000088247 Q92945                KHSRP</span></span>
<span><span class="co">## 2 ENSG00000162613 Q96AE4                FUBP1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="format">Format<a class="anchor" aria-label="anchor" href="#format"></a>
</h3>
<p>The UniProt REST API can deliver results in different data formats.
The formats available depends on the database being accessed and the
<strong>uniprotREST</strong> function being used. See the X page for a
full list of available formats.</p>
<p>The <strong>uniprotREST</strong> wrapper functions do not support all
formats yet. Each tool currently supports the following formats:</p>
<ul>
<li>
<code><a href="../reference/uniprot_map.html">uniprot_map()</a></code> and <code><a href="../reference/uniprot_search.html">uniprot_search()</a></code> =
<code>tsv, fasta</code>
</li>
<li>
<code><a href="../reference/uniprot_single.html">uniprot_single()</a></code> = <code>tsv, fasta, json</code>
</li>
</ul>
<p>Here we’ll re-use the <code>result3</code> job above, but request the
FASTA protein sequences instead. If the <code>Biostrings</code> package
is installed (highly recommended) the output will be a
<code><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">Biostrings::AAStringSet</a></code>, or otherwise a
<code>named character</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>result4 <span class="ot">&lt;-</span> <span class="fu">uniprot_map</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">ids =</span> goi,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"Ensembl"</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"UniProtKB-Swiss-Prot"</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">"fasta"</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Running job: 03772c110ee17372f3dc29049d1ff9fa50105fad </span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a> Downloading<span class="sc">:</span> page <span class="dv">1</span> of <span class="dv">1</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result4</span></span>
<span><span class="co">## AAStringSet object of length 2:</span></span>
<span><span class="co">##     width seq                                               names               </span></span>
<span><span class="co">## [1]   711 MSDYSTGGPPPGPPPPAGGGGGA...YGQTPGPGGPQPPPTQQGQQQAQ sp|Q92945|FUBP2_H...</span></span>
<span><span class="co">## [2]   644 MADYSTVPPPSSGSAGGGGGGGG...QAAYYAQTSPQGMPQHPPAPQGQ sp|Q96AE4|FUBP1_H...</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="path">Path<a class="anchor" aria-label="anchor" href="#path"></a>
</h3>
<p>The previous examples all save the data from UniProt into an object
in memory. However, you can also save the data to a file on disk. To do
this, just specify a file path with the correct extension. The file must
not already exist otherwise an error is thrown.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get temp path for this example (and delete when done)</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">".tsv"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">on.exit</span>(<span class="fu">unlink</span>(tmp))</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Save results to a tsv file</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">uniprot_map</span>(</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">ids =</span> goi,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"Ensembl"</span>,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"UniProtKB-Swiss-Prot"</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">fields =</span> <span class="fu">c</span>(<span class="st">"accession"</span>, <span class="st">"gene_primary"</span>),</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">"tsv"</span>,</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> tmp</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="do">## Running job: 03772c110ee17372f3dc29049d1ff9fa50105fad </span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a> Downloading<span class="sc">:</span> page <span class="dv">1</span> of <span class="dv">1</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Check file contents</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="fu">read.delim</span>(tmp)</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a><span class="do">##              From  Entry Gene.Names..primary.</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 ENSG00000088247 Q92945                KHSRP</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 ENSG00000162613 Q96AE4                FUBP1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="other-arguments">Other arguments<a class="anchor" aria-label="anchor" href="#other-arguments"></a>
</h3>
<p>The other arguments in <code><a href="../reference/uniprot_map.html">uniprot_map()</a></code> are as follows:</p>
<div class="section level4">
<h4 id="isoform">Isoform<a class="anchor" aria-label="anchor" href="#isoform"></a>
</h4>
<p>By default, the UniProt APIs will only provide results with a
proteins’ <a href="https://www.uniprot.org/help/canonical_and_isoforms" class="external-link">canonical
sequence</a>. If you set <code>isoform = TRUE</code>, then isoform
sequences will be included as well. This is typically only relevant when
<code>format = "fasta"</code> although I have run into some
exceptions.</p>
<p>Here we get the canonical and isoform sequence for human GAPDH.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>result5 <span class="ot">&lt;-</span> <span class="fu">uniprot_map</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">ids =</span> <span class="st">"P04406"</span>,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">"fasta"</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">isoform =</span> <span class="cn">TRUE</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Running job: 8589ca2e878ae84053759d9161aabd9006a79c6d </span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a> Downloading<span class="sc">:</span> page <span class="dv">1</span> of <span class="dv">1</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result5</span></span>
<span><span class="co">## AAStringSet object of length 2:</span></span>
<span><span class="co">##     width seq                                               names               </span></span>
<span><span class="co">## [1]   335 MGKVKVGVNGFGRIGRLVTRAAF...WYDNEFGYSNRVVDLMAHMASKE sp|P04406|G3P_HUM...</span></span>
<span><span class="co">## [2]   293 MVYMFQYDSTHGKFHGTVKAENG...WYDNEFGYSNRVVDLMAHMASKE sp|P04406-2|G3P_H...</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="method-and-page_size">Method and page_size<a class="anchor" aria-label="anchor" href="#method-and-page_size"></a>
</h4>
<p>The UniProt API provides results via 2 endpoints: stream, and
pagination, which you can choose via the <code>method</code>
argument.</p>
<p>By default, <code><a href="../reference/uniprot_map.html">uniprot_map()</a></code> and
<code><a href="../reference/uniprot_search.html">uniprot_search()</a></code> use <code>method = "paged"</code> which is
more robust but slightly slower, with the default recommended
<code>page_size</code> of 500. Whereas <code><a href="../reference/uniprot_single.html">uniprot_single()</a></code>
only uses the stream endpoint.</p>
<p>Paged endpoint:</p>
<ul>
<li>Slightly slower.</li>
<li>Processes results in chunks, so much more reliable to connection
issues.</li>
<li>Can theoretically handle more than 10,000,000 results.</li>
</ul>
<p>Stream endpoint:</p>
<ul>
<li>Slightly faster.</li>
<li>Expensive for the API, uses a lot of memory.</li>
<li>Can return a <code>429</code> status error if it currently has too
many requests.</li>
<li>Up to 10,000,000 results can be fetched.</li>
</ul>
</div>
<div class="section level4">
<h4 id="compressed">Compressed<a class="anchor" aria-label="anchor" href="#compressed"></a>
</h4>
<p>Should gzipped data be requested? This is <code>FALSE</code> by
default, and it is only used if <code>method = "stream"</code> and
<code>path</code> is specified.</p>
<p>For example:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get temp path for this example (and delete when done)</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">tempfile</span>(<span class="at">fileext =</span> <span class="st">".fasta.gz"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">on.exit</span>(<span class="fu">unlink</span>(tmp))</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Save results to a tsv file</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="fu">uniprot_map</span>(</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">ids =</span> <span class="st">"P04406"</span>,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">"fasta"</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">isoform =</span> <span class="cn">TRUE</span>,</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"stream"</span>,</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">path =</span> tmp,</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">compressed =</span> <span class="cn">TRUE</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="do">## Running job: 8589ca2e878ae84053759d9161aabd9006a79c6d </span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">0</span> B     </span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">0</span> B     </span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">0</span> B     </span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">0</span> B     </span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">0</span> B     </span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">0</span> B     </span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">0</span> B     </span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">0</span> B     </span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">420</span> B     </span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">420</span> B     </span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">420</span> B     </span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>Downloading<span class="sc">:</span> <span class="dv">420</span> B</span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Check file contents</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">readAAStringSet</span>(tmp)</span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a><span class="do">## AAStringSet object of length 2:</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a><span class="do">##     width seq                                               names               </span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a><span class="do">## [1]   335 MGKVKVGVNGFGRIGRLVTRAAF...WYDNEFGYSNRVVDLMAHMASKE sp|P04406|G3P_HUM...</span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="do">## [2]   293 MVYMFQYDSTHGKFHGTVKAENG...WYDNEFGYSNRVVDLMAHMASKE sp|P04406-2|G3P_H...</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="verbosity">Verbosity<a class="anchor" aria-label="anchor" href="#verbosity"></a>
</h4>
<p>Controls the amount of information to print:</p>
<ul>
<li>Use <code>verbosity = 0</code> to not print anything.</li>
<li>Use <code>verbosity = 1</code>, <code>2</code>, or <code>3</code> to
print increasing amounts of information about the HTTP requests made to
the UniProt API (typically for debugging purposes).</li>
</ul>
</div>
<div class="section level4">
<h4 id="dry_run">Dry_run<a class="anchor" aria-label="anchor" href="#dry_run"></a>
</h4>
<p>If <code>TRUE</code>, performs the request locally with
<code><a href="https://httr2.r-lib.org/reference/req_dry_run.html" class="external-link">httr2::req_dry_run()</a></code> instead of the actually sending it to
the UniProt REST API. This is useful for debugging purposes if you are
getting <code>400 - Bad request</code> status errors.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="querying-uniprot-with-uniprot_search">2. Querying UniProt with <code>uniprot_search</code><a class="anchor" aria-label="anchor" href="#querying-uniprot-with-uniprot_search"></a>
</h2>
<p>This function is used to perform text searches against UniProt, akin
to using the search bar on their website. The different databases
available from the search bar are also available via
<code><a href="../reference/uniprot_search.html">uniprot_search()</a></code> (see page x).</p>
<p>It’s very important that the search string is constructed correctly,
see <a href="https://www.uniprot.org/help/text-search" class="external-link">this page</a> for
help building queries. If you get a <code>400 - Bad request</code>
error, its likely your search string is not formatted correctly.</p>
<p>Here we’ll do a search for human proteins annotated with the
glycoprotein keyword, which are in SwissProt i.e. have been manually
reviewed.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>result6 <span class="ot">&lt;-</span> <span class="fu">uniprot_search</span>(</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">query =</span> <span class="st">"(proteome:UP000005640) AND (keyword:KW-0325) AND (length&lt;100)"</span>,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">database =</span> <span class="st">"uniprotkb"</span>,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">"tsv"</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">fields =</span> <span class="fu">c</span>(<span class="st">"accession"</span>, <span class="st">"gene_primary"</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a> Downloading<span class="sc">:</span> page <span class="dv">1</span> of <span class="dv">1</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result6</span><span class="op">)</span></span>
<span><span class="co">##    Entry Gene.Names..primary.</span></span>
<span><span class="co">## 1 P06028                 GYPB</span></span>
<span><span class="co">## 2 P80098                 CCL7</span></span>
<span><span class="co">## 3 Q16627                CCL14</span></span>
<span><span class="co">## 4 P0DMC3                APELA</span></span>
<span><span class="co">## 5 P25063                 CD24</span></span>
<span><span class="co">## 6 P31358                 CD52</span></span></code></pre></div>
<p>The other UniProt databases other than UniProtKB are available to
query as well. In this example we’ll look for all reference proteomes
with the word ‘dog’ in their title.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>result7 <span class="ot">&lt;-</span> <span class="fu">uniprot_search</span>(</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"dog"</span>,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">database =</span> <span class="st">"proteomes"</span>,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">"tsv"</span>,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">fields =</span> <span class="fu">c</span>(<span class="st">"upid"</span>, <span class="st">"organism"</span>)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a> Downloading<span class="sc">:</span> page <span class="dv">1</span> of <span class="dv">1</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result7</span><span class="op">)</span></span>
<span><span class="co">##   Proteome.Id</span></span>
<span><span class="co">## 1 UP000252519</span></span>
<span><span class="co">## 2 UP000645828</span></span>
<span><span class="co">## 3 UP000002254</span></span>
<span><span class="co">## 4 UP000029752</span></span>
<span><span class="co">## 5 UP000201396</span></span>
<span><span class="co">## 6 UP000792793</span></span>
<span><span class="co">##                                                                                            Organism</span></span>
<span><span class="co">## 1                                                                Ancylostoma caninum (Dog hookworm)</span></span>
<span><span class="co">## 2                                       Nyctereutes procyonoides (Raccoon dog) (Canis procyonoides)</span></span>
<span><span class="co">## 3                                                   Canis lupus familiaris (Dog) (Canis familiaris)</span></span>
<span><span class="co">## 4 Cadicivirus A (isolate Dog/Hong Kong/209/2008) (CaPdV-1) (Canine picodicistrovirus (isolate 209))</span></span>
<span><span class="co">## 5                                                                             Raccoon dog amdovirus</span></span>
<span><span class="co">## 6                                               Lycaon pictus (African wild dog) (Cape hunting dog)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="retrieving-an-entry-with-uniprot_single">3. Retrieving an entry with <code>uniprot_single</code><a class="anchor" aria-label="anchor" href="#retrieving-an-entry-with-uniprot_single"></a>
</h2>
<p><code><a href="../reference/uniprot_single.html">uniprot_single()</a></code> is used to quickly retrieve information
about a single entry in UniProt. By default the <code>json</code> format
is requested and is parsed into a list which contains <em>all</em>
information available for that particular entry. All other arguments
work the same as <code><a href="../reference/uniprot_search.html">uniprot_search()</a></code>.</p>
<p>For example:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result8</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/uniprot_single.html">uniprot_single</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="st">"P99999"</span>,</span>
<span>  verbosity <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">result8</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## List of 17</span></span>
<span><span class="co">##  $ entryType               : chr "UniProtKB reviewed (Swiss-Prot)"</span></span>
<span><span class="co">##  $ primaryAccession        : chr "P99999"</span></span>
<span><span class="co">##  $ secondaryAccessions     :List of 6</span></span>
<span><span class="co">##  $ uniProtkbId             : chr "CYC_HUMAN"</span></span>
<span><span class="co">##  $ entryAudit              :List of 5</span></span>
<span><span class="co">##  $ annotationScore         : num 5</span></span>
<span><span class="co">##  $ organism                :List of 4</span></span>
<span><span class="co">##  $ proteinExistence        : chr "1: Evidence at protein level"</span></span>
<span><span class="co">##  $ proteinDescription      :List of 1</span></span>
<span><span class="co">##  $ genes                   :List of 1</span></span>
<span><span class="co">##  $ comments                :List of 10</span></span>
<span><span class="co">##  $ features                :List of 36</span></span>
<span><span class="co">##  $ keywords                :List of 14</span></span>
<span><span class="co">##  $ references              :List of 19</span></span>
<span><span class="co">##  $ uniProtKBCrossReferences:List of 178</span></span>
<span><span class="co">##  $ sequence                :List of 5</span></span>
<span><span class="co">##  $ extraAttributes         :List of 3</span></span></code></pre></div>
<p>Again, there are other UniProt databases available apart from
UniProtKB (see X page for a full list of databases).</p>
<p>For example UniParc:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result9</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/uniprot_single.html">uniprot_single</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="st">"UPI0001C61C61"</span>,</span>
<span>  database <span class="op">=</span> <span class="st">"uniparc"</span>,</span>
<span>  verbosity <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">result9</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## List of 6</span></span>
<span><span class="co">##  $ uniParcId                : chr "UPI0001C61C61"</span></span>
<span><span class="co">##  $ uniParcCrossReferences   :List of 4</span></span>
<span><span class="co">##  $ sequence                 :List of 5</span></span>
<span><span class="co">##  $ sequenceFeatures         :List of 7</span></span>
<span><span class="co">##  $ oldestCrossRefCreated    : chr "2010-03-03"</span></span>
<span><span class="co">##  $ mostRecentCrossRefUpdated: chr "2022-12-14"</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Charlotte Dawson.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
